moped-editor/src/views/projects/projectView/ProjectMilestones.js

```diff
@@ -0,0 +1,368 @@
+import React from "react";
+
+// Material
+import {
+  Button,
+  CircularProgress,
+  Grid,
+  TextField,
+  Switch,
+  Typography,
+  FormControl,
+  FormHelperText,
+} from "@material-ui/core";
+import {
+  AddCircle as AddCircleIcon,
+  EditOutlined as EditOutlinedIcon,
+} from "@material-ui/icons";
+import MaterialTable, {
+  MTableEditRow,
+  MTableAction,
+} from "@material-table/core";
+import { handleKeyEvent } from "../../../utils/materialTableHelpers";
+import typography from "../../../theme/typography";
+
+// Query
+import {
+  UPDATE_PROJECT_MILESTONES_MUTATION,
+  DELETE_PROJECT_MILESTONE,
+  ADD_PROJECT_MILESTONE,
+} from "../../../queries/project";
+import { PAGING_DEFAULT_COUNT } from "../../../constants/tables";
+import { useMutation } from "@apollo/client";
+import { format } from "date-fns";
+import parseISO from "date-fns/parseISO";
+import Autocomplete from "@material-ui/lab/Autocomplete";
+
+/**
+ * ProjectTimeline Component - renders the view displayed when the "Timeline"
+ * tab is active
+ * @return {JSX.Element}
+ * @constructor
+ */
+const ProjectMilestones = ({ projectId, loading, data, refetch }) => {
+  /** addAction Ref - mutable ref object used to access add action button
+   * imperatively.
+   * @type {object} addActionRef
+   * */
+  const addActionRefMilestones = React.useRef();
+
+  // Mutations
+  const [updateProjectMilestone] = useMutation(
+    UPDATE_PROJECT_MILESTONES_MUTATION
+  );
+  const [deleteProjectMilestone] = useMutation(DELETE_PROJECT_MILESTONE);
+  const [addProjectMilestone] = useMutation(ADD_PROJECT_MILESTONE);
+
+  // If the query is loading or data object is undefined,
+  // stop here and just render the spinner.
+  if (loading || !data) return <CircularProgress />;
+
+  /**
+   * Phase table lookup object formatted into the shape that Dropdown expects
+   * Ex: { 1: "Potential", 2: "Planned", ...}
+   */
+  const phaseNameLookup = data.moped_phases.reduce(
+    (obj, item) =>
+      Object.assign(obj, {
+        [item.phase_id]: item.phase_name,
+      }),
+    {}
+  );
+
+  /**
+   * Milestone table lookup object formatted into the shape that <Autocomplete> expects.
+   * Ex: { 1: "Award", 2: "Bid", ...}
+   */
+  const milestoneNameLookup = data.moped_milestones.reduce(
+    (obj, item) =>
+      Object.assign(obj, {
+        [item.milestone_id]: item.milestone_name,
+      }),
+    {}
+  );
+
+  /**
+   * DateFieldEditComponent - renders a Date type Calendar select
+   * @param {object} props - Values passed through Material Table `editComponent`
+   * @return {JSX.Element}
+   * @constructor
+   */
+  const DateFieldEditComponent = (props) => (
+    <TextField
+      name={props.name}
+      label={props.label}
+      type="date"
+      variant="standard"
+      value={props.value}
+      onChange={(e) => props.onChange(e.target.value)}
+      onKeyDown={(e) => handleKeyEvent(e)}
+      InputLabelProps={{
+        shrink: true,
+      }}
+    />
+  );
```

Comment:
for **components** that are identical in both phases and milestones, would prefer to see individual component modules in this directory. `DateFieldEditComponent.js`, `ToggleEditComponent.js`, etc.